feat: update finetune script to support Liger kernels

iharshulhan opened the pull request:
This pull request updates multiple fine-tuning components in the `assets/training/finetune_acft_hf_nlp` directory to add support for the [Liger](https://github.com/linkedin/Liger-Kernel) kernel, which significantly optimizes memory consumption for long sequence fine-tuning. 

### Version Updates:
* Incremented version numbers from `0.0.75` to `0.0.76` for all components: `chat_completion`, `question_answering`, `summarization`, and `text_classification`. [[1]](diffhunk://#diff-f9fa561549383e2a90e37a6069acb9cc96e48cd38baff230ff5665a1c8b346e7L3-R3) [[2]](diffhunk://#diff-cde8f3340ad1b54092c215a8c5e5a1143a48c4b3861a584eb919f77e664b4c84L3-R3) [[3]](diffhunk://#diff-9538c4682d047a44fe07a674d6d94b8bb0a838e0762a82e37611745fa0cdd14cL3-R3) [[4]](diffhunk://#diff-3a4b9ac7c8deec69a6812fe4be494aa6b24f45a38aecb5dbd6b8ba643e669aaeL3-R3)

### Feature Enhancements:
* Added a new input parameter `use_liger_kernel` to all components. This parameter enables the Liger kernel for optimized attention and defaults to `"false"`. [[1]](diffhunk://#diff-f9fa561549383e2a90e37a6069acb9cc96e48cd38baff230ff5665a1c8b346e7R57-R65) [[2]](diffhunk://#diff-cde8f3340ad1b54092c215a8c5e5a1143a48c4b3861a584eb919f77e664b4c84R57-R65) [[3]](diffhunk://#diff-9538c4682d047a44fe07a674d6d94b8bb0a838e0762a82e37611745fa0cdd14cR57-R65) [[4]](diffhunk://#diff-3a4b9ac7c8deec69a6812fe4be494aa6b24f45a38aecb5dbd6b8ba643e669aaeR57-R65)
* Updated the `command` section in each component to include the `use_liger_kernel` parameter in the Python script execution. [[1]](diffhunk://#diff-f9fa561549383e2a90e37a6069acb9cc96e48cd38baff230ff5665a1c8b346e7L351-R359) [[2]](diffhunk://#diff-cde8f3340ad1b54092c215a8c5e5a1143a48c4b3861a584eb919f77e664b4c84L350-R359) [[3]](diffhunk://#diff-9538c4682d047a44fe07a674d6d94b8bb0a838e0762a82e37611745fa0cdd14cL352-R361) Fa2dc819L349R358)


###  Benchmark Results

Here are some internal benchmark numbers for Llama-3.2-3B-Instruct fine-tune run on A100 GPU.
Memory is measured as peak memory recorded per GPU in GB. Total time is time for fine-tuning for 1 epoch with 60 samples.

| Liger kernel | DeepSpeed stage       | Sequence length | GPUs | Train time (s) | GPU memory (GB) | Batch size |
|:-------------|:----------------------|----------------:|-----:|---------------:|---------------:|-----------:|
| False        | off                   | 30000           | 1    | 593            | 68             | 1          |
| False        | off                   | 30000           | 2    | 335            | 68             | 1          |
| False        | Stage 1               | 30000           | 2    | 339            | 68             | 1          |
| False        | Stage 2               | 30000           | 2    | 340            | 68             | 1          |
| False        | Stage 3               | 30000           | 2    | 350            | 70             | 1          |
| False        | Stage 3 + offload     | 30000           | 2    | 342            | 68             | 1          |
| True         | off                   | 30000           | 1    | 531            | 18             | 1          |
| True         | off                   | 30000           | 2    | 303            | 18             | 1          |
| True         | off                   | 80000           | 1    | 2276           | 40             | 1          |
| True         | off                   | 80000           | 2    | 1231           | 41             | 1          |
| True         | Stage 1               | 30000           | 2    | 304            | 19             | 1          |
| True         | Stage 1               | 80000           | 2    | 1238           | 38             | 1          |
| True         | Stage 2               | 30000           | 2    | 308            | 19             | 1          |
| True         | Stage 2               | 80000           | 2    | 1239           | 39             | 1          |
| True         | Stage 3               | 30000           | 2    | 306            | 20             | 1          |
| True         | Stage 3               | 80000           | 2    | 1254           | 41             | 1          |
| True         | Stage 3 + offload     | 30000           | 2    | 314            | 19             | 1          |
| True         | Stage 3 + offload     | 80000           | 2    | 1272           | 39             | 1          |

Key observations:
- The Liger attention kernel dramatically reduces per-GPU memory usage (e.g., down to ~18 GB vs. ~68 GB at 30 K sequence length) while maintaining or improving throughput.

babu-namburi commented on assets/training/finetune_acft_hf_nlp/components/finetune/chat_completion/spec.yaml:
```diff
@@ -213,7 +222,6 @@ inputs:
     enum:
     - epoch
     - steps
-    - disable
```
Why was this removed ?

iharshulhan replied:
You're right, this should be reverted. Should we add to other components though? Like text_classification, question_answering, etc.

babu-namburi commented on assets/training/finetune_acft_hf_nlp/components/finetune/chat_completion/spec.yaml:
```diff
@@ -1,6 +1,6 @@
 $schema: https://azuremlschemas.azureedge.net/latest/commandComponent.schema.json
 name: chat_completion_finetune
-version: 0.0.75
```
To make it easier to update versions in the control tower and Azure AI studio, we consistently bump all the component and pipeline versions.

iharshulhan replied:
@babu-namburi The versions were bumped across all components and pipelines

babu-namburi commented on assets/training/finetune_acft_hf_nlp/src/finetune/finetune.py:
```diff
@@ -1104,6 +1111,9 @@ def finetune(args: Namespace):
     model_asset_id = getattr(args, "model_asset_id", None) or ""
 
+    eval_strategy = getattr(args, "eval_strategy", "steps")
+    setattr(args, "eval_strategy", eval_strategy)
```
The eval_strategy is set in line:1293

iharshulhan commented:
## Changes update

This PR updates the version numbers and environment references across various pipeline and component specifications while adding a new flag to enable the Liger kernel for optimized attention in finetuning components.  
- Updated component versions from 0.0.76 to 0.0.77 (and pipelines from 0.0.77 to 0.0.78)  
- Updated environment versions from 99 to 100 where applicable  
- Added a new input flag "use_liger_kernel" in several finetune specs to control optimized attention

| File | Description |
| ---- | ----------- |
| assets/training/finetune_acft_hf_nlp/components/pipeline_components/*/spec.yaml | Version bump and component dependency update for pipelines |
| assets/training/finetune_acft_hf_nlp/components/model_import/*/spec.yaml | Version and environment update for model import components |
| assets/training/finetune_acft_hf_nlp/components/model_converter/common/spec.yaml | Version and environment update with a minor command formatting change |
| assets/training/finetune_acft_hf_nlp/components/finetune/*/spec.yaml | Version and environment update; addition of "use_liger_kernel" flag and corresponding command-line updates |
